BLUESPAWN-common/headers/common/wrappers.hpp

```diff
@@ -0,0 +1,328 @@
+#pragma once
+#include <Windows.h>
+
+#include <map>
+#include <string>
+#include <memory>
+#include <optional>
+#include <functional>
+
+template<class T>
+class GenericWrapper {
+protected:
+	static inline std::map<T, DWORD> mReferenceCounts{};
+
+	T WrappedObject;
+	T BadValue;
+	bool bFreeOnDestruction = true;
+
+	std::function<void(T)> freeResource;
+
+	void DestroyReference(){
+		mReferenceCounts[WrappedObject]--;
+		if(mReferenceCounts[WrappedObject] == 0){
+			mReferenceCounts.erase(WrappedObject);
+			if(bFreeOnDestruction && operator bool()){
+				freeResource(WrappedObject);
+			}
+		}
+		WrappedObject = BadValue;
+	}
+
+	void SetReference(T object){
+		WrappedObject = object;
+		if(mReferenceCounts.find(object) == mReferenceCounts.end()){
+			mReferenceCounts.emplace(object, 1);
+		} else {
+			mReferenceCounts[object]++;
+		}
+	}
+
+public:
+
+	GenericWrapper(T object, std::function<void(T)> freeFunction = [](T object){ delete object; }, T BadValue = nullptr)
+		: WrappedObject{ object }, freeResource{ freeFunction }, BadValue{ BadValue } { SetReference(object); }
+
+	GenericWrapper(const GenericWrapper& copy)
+		: freeResource{ copy.freeResource } { SetReference(copy.WrappedObject); }
+
+	GenericWrapper(GenericWrapper&& move)
+		: freeResource{ move.freeResource } { SetReference(move.WrappedObject); move.DestroyReference(); }
+
+	GenericWrapper& operator=(const GenericWrapper& copy){ 
+		freeResource = copy.freeResource; 
+		DestroyReference();
+		SetReference(copy.WrappedObject); 
+		return *this;
+	}
+	GenericWrapper&& operator=(GenericWrapper&& move) = delete;
+
+	~GenericWrapper(){ DestroyReference(); }
+
+	operator T() const { return WrappedObject; }
+	T* operator *(){ return *WrappedObject; }
+	T& operator &(){ return &WrappedObject; }
+	bool operator ==(T object){ return WrappedObject == object; }
+	bool operator !(){ return !WrappedObject || WrappedObject == BadValue; }
+	operator bool(){ return !operator!(); }
+
+	T Get() const { return WrappedObject; }
+	bool Release(){ return bFreeOnDestruction != (bFreeOnDestruction = false); }
+	bool Lock(){ return bFreeOnDestruction != (bFreeOnDestruction = true);  }
+	DWORD GetReferenceCount(){ return mReferenceCounts[WrappedObject]; }
+};
+
+class HandleWrapper : public GenericWrapper<HANDLE> {
+public:
+	HandleWrapper(HANDLE handle) :
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Class 'HandleWrapper' has a constructor with 1 argument that is not explicit.](https://app.codacy.com/manual/ION28/BLUESPAWN/pullRequest?prid=4994516)